packages/payment-processor/src/payment/erc20-transferable-receivable.ts

```diff
@@ -0,0 +1,250 @@
+  signerOrProvider: providers.Provider | Signer = getProvider(),
+  amount?: BigNumberish,
+  feeAmount?: BigNumberish,
+  overrides?: ITransactionOverrides,
+): Promise<ContractTransaction> {
+  const { data, to, value } = await prepareErc20TransferableReceivablePaymentTransaction(
+    request,
+    signerOrProvider,
+    amount,
+    feeAmount,
+  );
+  const signer = getSigner(signerOrProvider);
+  return signer.sendTransaction({ data, to, value, ...overrides });
+}
+
+/**
+ * Encodes the call to pay a request through the ERC20 receivable contract, can be used with a Multisig contract.
+ * @param request request to pay
+ * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
+ * @param amount optionally, the amount to pay. Defaults to remaining amount of the request.
+ * @param feeAmountOverride optionally, the fee amount to pay. Defaults to the fee amount of the request.
+ */
+export async function prepareErc20TransferableReceivablePaymentTransaction(
+  request: ClientTypes.IRequestData,
+  signerOrProvider: providers.Provider | Signer,
+  amount?: BigNumberish,
+  feeAmountOverride?: BigNumberish,
+): Promise<IPreparedTransaction> {
+  validateRequest(request, ExtensionTypes.PAYMENT_NETWORK_ID.ERC20_TRANSFERABLE_RECEIVABLE);
+
+  return {
+    data: await encodePayErc20TransferableReceivableRequest(
+      request,
+      signerOrProvider,
+      amount,
+      feeAmountOverride,
+    ),
+    to: getProxyAddress(
+      request,
+      Erc20PaymentNetwork.ERC20TransferableReceivablePaymentDetector.getDeploymentInformation,
+    ),
+    value: 0,
+  };
+}
+
+/**
+ * Encodes the call to pay a request through the ERC20 receivable contract, can be used with a Multisig contract.
+ * @param request request to pay
+ * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
+ * @param amount optionally, the amount to pay. Defaults to remaining amount of the request.
+ * @param feeAmountOverride optionally, the fee amount to pay. Defaults to the fee amount of the request.
+ */
+export async function encodePayErc20TransferableReceivableRequest(
+  request: ClientTypes.IRequestData,
+  signerOrProvider: providers.Provider | Signer,
+  amount?: BigNumberish,
+  feeAmountOverride?: BigNumberish,
+): Promise<string> {
+  validateRequest(request, ExtensionTypes.PAYMENT_NETWORK_ID.ERC20_TRANSFERABLE_RECEIVABLE);
+
+  const amountToPay = getAmountToPay(request, amount);
+  const { paymentReference, feeAddress, feeAmount } = getRequestPaymentValues(request);
+  const feeToPay = BigNumber.from(feeAmountOverride || feeAmount || 0);
+
+  const receivableContract = ERC20TransferableReceivable__factory.createInterface();
+
+  const receivableTokenId = await getReceivableTokenIdForRequest(request, signerOrProvider);
+
+  if (receivableTokenId.isZero()) {
+    throw new Error(
+      'The receivable for this request has not been minted yet. Please check with the payee.',
+    );
+  }
```

Comment:
This validation logic seems like it should be moved to a new `validatePayErc20TransferableReceivable` function located in `packages/payment-processor/src/payment/utils.ts` so that it matches the existing code architecture. Additionally, I think `validatePayErc20TransferableReceivable` should call `validateErc20FeeProxyRequest` so that request and fee get validated as well.

The following functions should call `validatePayErc20TransferableReceivable`:

* `prepareErc20TransferableReceivablePaymentTransaction`
* `encodePayErc20TransferableReceivableRequest`
* `_getErc20TransferableReceivablePaymentUrl` (even though `_getErc20FeeProxyPaymentUrl` calls `validateRequest`. I think it's wrong.)

Reply:
I just realized that `getReceivableTokenIdForRequest` makes an RPC to the Ethereum node provider. I don't think we should RPC in both `prepare...` and `encode...`. So instead, `validatePayErc20TransferableReceivable should be called from `payErc20TransferableReceivableRequest`. I recognize this diverges from the existing architecture. I edited my previous comment to reflect this.

Reply:
Per other [comment](https://github.com/RequestNetwork/requestNetwork/pull/1033#discussion_r1105018198) the `_getErc20TransferableReceivablePaymentUrl` function is removed. I updated my initial comment in this thread to reflect this.